### PR TITLE
refactor: convert config actions to TypeScript

### PR DESCRIPTION
**Pull request by smashercosmo**
**Summary**

Convert config actions to TypeScript

**Test plan**

Run current tests

### CODE REVIEW

**Comment by smashercosmo** [comment]
@erezrokah we have to decide how to avoid types duplication. Currently I just copied types from index.d.ts to types/redux.ts, because I couldn't use them directly from index.d.ts, but probably we should reexport them from redux.ts to index.d.ts.

**Comment by smashercosmo** [comment]
Seems like failing tests are not related to this PR. They are failing also in master.

**Comment by erezrokah** [comment]
> but probably we should reexport them from redux.ts to index.d.ts.

Agree, we had `index.d.ts` before we had any TypeScript in the repo.

**Comment by smashercosmo** [comment]
Hm... I can see many failed tests related to multiple_folders mode, but according to the code, this mode is currently not supported. @erezrokah am I missing something?

**Comment by erezrokah** [comment]
> I can see many failed tests related to multiple_folders mode

Multiple folders is (should be) supported only for folder collections

**Comment by smashercosmo** [comment]
Oh, ok. My mistake then.

**Comment by smashercosmo** [comment]
> but probably we should reexport them from redux.ts to index.d.ts.

should I do it in this PR?

**Comment by erezrokah** [comment]
> should I do it in this PR?

I think we can do it in another one to make it easier to track the changes.

File: packages/netlify-cms-core/src/actions/config.ts [hunk 1]
```diff
@@ -376,12 +376,13 @@ export function parseConfig(data: string) {
 }
 
 async function getConfigYaml(file: string, hasManualConfig: boolean) {
-  const response = await fetch(file, { credentials: 'same-origin' });
-  if (!response.ok || response.status !== 200) {
```
Comment (erezrokah) [inline packages/netlify-cms-core/src/actions/config.ts]: `!response.ok` already means `response.status !== 200`. We are looking for `200` status code.
Also `fetch` can throw so we would like to handle that case as well

File: packages/netlify-cms-core/src/actions/config.ts [hunk 2]
```diff
@@ -56,17 +56,16 @@ function traverseFieldsJS<Field extends CmsField>(
...
+  const validTypes: { [type: string]: string } = {
     'text/yaml': 'yaml',
     'application/x-yaml': 'yaml',
   };
   const configLinkEl = document.querySelector<HTMLLinkElement>('link[rel="cms-config-url"]');
-  const configLinkType = configLinkEl && configLinkEl.getAttribute('type');
-  const configLinkHref = configLinkEl && configLinkEl.getAttribute('href');
-  const configUrl =
-    configLinkType && configLinkHref && validTypes[configLinkType] ? configLinkHref : 'config.yml';
-  console.log(`Using config file path: "${configUrl}"`);
```
Comment (erezrokah) [inline packages/netlify-cms-core/src/actions/config.ts]: This log should be printed only when using a custom config URL

File: packages/netlify-cms-core/src/actions/config.ts [hunk 3]
```diff
@@ -386,10 +386,10 @@ async function getConfigYaml(file: string, hasManualConfig: boolean) {
   const contentType = response.headers.get('Content-Type') || 'Not-Found';
   const isYaml = contentType.indexOf('yaml') !== -1;
   if (!isYaml) {
+    console.log(`Response for ${file} was not yaml. (Content-Type: ${contentType})`);
     if (hasManualConfig) {
       return {};
     }
-    throw new Error(`Response for ${file} was not yaml. (Content-Type: ${contentType})`);
```
Comment (erezrokah) [inline packages/netlify-cms-core/src/actions/config.ts]: We still want to attempt parse the configuration in this case.
For example, when you deploy to S3 via the `aws-cli` it sometimes fails to set the correct mime type for files. See https://github.com/aws/aws-cli/issues/1249

**Comment by erezrokah** [comment]
@smashercosmo I think I forced pushed over some of your fixes 🤦

File: packages/netlify-cms-core/src/actions/config.ts (hunk 2 above)
Comment (smashercosmo) [reply packages/netlify-cms-core/src/actions/config.ts]: Oh, ok

File: packages/netlify-cms-core/src/actions/config.ts (hunk 3 above)
Comment (smashercosmo) [reply packages/netlify-cms-core/src/actions/config.ts]: Alright, got it

File: packages/netlify-cms-core/src/actions/config.ts (hunk 1 above)
Comment (smashercosmo) [reply packages/netlify-cms-core/src/actions/config.ts]: Could you give an example of fetch throwing? Maybe this should be covered by unit test then.

**Comment by smashercosmo** [comment]
@erezrokah you need to rollback changes to config.spec.ts. They are not needed.

File: packages/netlify-cms-core/src/actions/config.ts (hunk 1 above)
Comment (erezrokah) [reply packages/netlify-cms-core/src/actions/config.ts]: `fetch('http://non-existing-host.com')` should throw.
I've added some tests to this logic in https://github.com/netlify/netlify-cms/pull/4950/commits/8150286cea83f9aed2720fbb85239c316ffd537e

**Review by erezrokah — APPROVED** [review]

File: packages/netlify-cms-core/src/actions/config.ts (hunk 1 above)
Comment (smashercosmo) [reply packages/netlify-cms-core/src/actions/config.ts]: Awesome 👍

**Pull request merged**